arch/x86/kvm/vmx/pkvm/hyp/ept.c

```diff
@@ -812,6 +854,17 @@ static bool allow_shadow_ept_mapping(struct pkvm_shadow_vm *vm,
 	if (vm->need_prepopulation) {
 		if (populate_pgstate_pgt(pgstate_pgt))
 			return false;
+		/*
+		 * Explicitly flush TLB of the host EPT after populating the page
+		 * state pgt.
+		 *
+		 * During the population, some pages are donated from primary VM to
+		 * this VM with the fastpath interface to avoid doing TLB flushing
+		 * during each iteration of the page donation so that to have a fast
+		 * population performance. So still need to do TLB flushing in the
+		 * end after finishing all the donations.
+		 */
+		host_ept_flush_tlb(&host_ept);
```

Comment:
Since pgstate table is used as an IOMMU page table, don't we also need to flush the CPU cache if the IOMMU is non-coherent? (AFAICS currently we don't do that at all, since `pgstate_pgt_mm_ops` has no `flush_cache` callback.)

Reply:
Yes, for non-coherent IOMMU, CPU cache should be flushed. The flush_cache callback for pgstate pgt should be added. Thanks for caching this.